tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala

```diff
@@ -1059,6 +1061,28 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
       reportError(s"Duplicate label named ${label.name}.")
   }
 
+  private def checkIsAsInstanceTargetType(refType: ReferenceType)(
+      implicit ctx: ErrorContext): Unit = {
+    refType match {
+      case ClassType(encodedName) =>
+        if (Definitions.isPrimitiveClass(encodedName)) {
+          reportError(
+              s"Primitive type $encodedName is not a valid target type for " +
+              "Is/AsInstanceOf")
+        } else {
+          val kind = lookupClass(encodedName).kind
+          if (kind.isJSType) {
+            reportError(
+                s"JS type $encodedName is not a valid target type for " +
+                "Is/AsInstanceOf")
+          }
+        }
+
+      case ArrayType(_, _) =>
```

Comment:
The reachability analysis does that, because the class data of any target type of a `Is/AsInstanceOf` is listed in the "referenced class data" fields of infos (which are checked by the info checker). If it's not found by the reachability analysis, you'll get a linking error before you even get to the IR checker.